### PR TITLE
Caching modules

### PR DESCRIPTION
**Pull request by HishamAlyahya**
This PR implements an automatic caching mechanism for the optimization of player modules. All compiled modules are now automatically cached to ~/.zse_cache/{lm_name}/{role}/{optimizer}/{dataset}/module.json

### CODE REVIEW

File: zero_sum_eval/caching.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from typing import Optional
+from pathlib import Path
+
+import os
+import dspy
+
+_DEFAULT_CACHE_DIR = os.path.join(Path().home(), ".zse_cache")
+
+def cache_module(module: dspy.Module, model: str, role: str, optimizer: str, dataset: str, cache_dir: Optional[str] = None) -> None:
```
Comment (haidark) [inline zero_sum_eval/caching.py:9]: can we call this save_module instead?

saving is more general than caching, as caching would be a special case.

File: zero_sum_eval/caching.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    if cache_dir is None:
+        cache_dir = _DEFAULT_CACHE_DIR
+    cache_dir = os.path.join(cache_dir, model, role, optimizer, dataset)
+    os.makedirs(cache_dir, exist_ok=True)
+
+    module.save(os.path.join(cache_dir, "module.json"))
+
+def load_cached_module(module: dspy.Module, model: str, role: str, optimizer: str, dataset: str, cache_dir: Optional[str] = None) -> Optional[dspy.Module]:
```
Comment (haidark) [inline zero_sum_eval/caching.py:17]: same here, can simply be load_module.

File: configs/chess.yaml [hunk 3]
```diff
@@ -25,7 +25,7 @@ game:
             dataset_args:
               filename: ./data/chess/stockfish_examples.jsonl
               role: White
-              num_examples: 10
+              num_samples: 10
```
Comment (haidark) [inline configs/chess.yaml:28]: does this change need to be made in all configs?

File: zero_sum_eval/games/chess/chess_dataset.py [hunk 4]
```diff
@@ -10,10 +10,11 @@
 
 @DATASET_REGISTRY.register("chess_dataset")
 class ChessDataset(Dataset):
-    def __init__(self, filename: str, role: Union[Literal["White"], Literal["Black"]]) -> None:
+    def __init__(self, filename: str, role: Union[Literal["White"], Literal["Black"]], num_samples: Optional[int] = None) -> None:
```
Comment (haidark) [inline zero_sum_eval/games/chess/chess_dataset.py:13]: can we pick one of samples/examples and use it consistently? I think samples may be better

Comment (haidark) [reply zero_sum_eval/games/chess/chess_dataset.py:13]: Anyways, it seems the changes in this file are unrelated.

File: configs/chess.yaml (hunk 3 above)
Comment (HishamAlyahya) [reply configs/chess.yaml]: This is only for this particular dataset class. We could have this as an argument for the super class though and have that functionality built-in to all datasets. Wdyt?

File: zero_sum_eval/games/chess/chess_dataset.py (hunk 4 above)
Comment (HishamAlyahya) [reply zero_sum_eval/games/chess/chess_dataset.py]: Examples is the term used in DSPy so I guess it makes more sense as you've done in the new commit.

File: configs/chess.yaml (hunk 3 above)
Comment (haidark) [reply configs/chess.yaml]: yeah, let's do that in the next iteration. I just reverted all the name changes and tested.

**Comment by haidark** [comment]
Tested with optimize: true for the following games

```
python run_game.py -c configs/chess.yaml 
python run_game.py -c configs/mathquiz.yaml
python run_game.py -c configs/gandalf.yaml 
```

**Pull request merged**